server/aesdsocket.c

```diff
@@ -186,58 +277,42 @@ int main(int argc, char *argv[]) {
         exit(EXIT_FAILURE);
     }
 
-    // Open the data file
-    tmp_file = fopen(DATA_FILE, "a+");
-    if (tmp_file == NULL) {
-        syslog(LOG_ERR, "Failed to open data file");
+    // Create the timer thread in the parent process.
+    if (pthread_create(&timer_thread_id, NULL, timer_thread, NULL) != 0) {
+        syslog(LOG_ERR, "Failed to create timer thread: %s", strerror(errno));
         cleanup();
         exit(EXIT_FAILURE);
     }
 
-    // Main server loop
+    // Main server loop: accept connections until stop_flag is set
     while (!stop_flag) {
         struct sockaddr_in client_addr;
-        socklen_t addr_len = sizeof client_addr;
-        int new_fd;
+        socklen_t addr_len = sizeof(client_addr);
+        int new_fd = accept(sockfd, (struct sockaddr *)&client_addr, &addr_len);
 
-        // Accept a new connection
-        if ((new_fd = accept(sockfd, (struct sockaddr *)&client_addr, &addr_len)) == -1) {
+        if (new_fd == -1) {
+            if (errno == EINTR) continue;
             syslog(LOG_ERR, "Failed to accept connection: %s", strerror(errno));
             continue;
         }
 
-        // Log the client's IP address and port
-        char client_ip[INET_ADDRSTRLEN];
-        inet_ntop(AF_INET, &client_addr.sin_addr, client_ip, sizeof client_ip);
-        int client_port = ntohs(client_addr.sin_port);
-        syslog(LOG_INFO, "Accepted connection from %s:%d", client_ip, client_port);
-
-        // Receive data from the client
-        char buf[BUF_SIZE];
-        ssize_t bytes_received;
-        while ((bytes_received = recv(new_fd, buf, BUF_SIZE - 1, 0)) > 0) {
-            buf[bytes_received] = '\0';
-            fputs(buf, tmp_file);
-            fflush(tmp_file);
-
-            // If a newline is found, send the file content back to the client
-            if (strchr(buf, '\n') != NULL) {
-                fseek(tmp_file, 0, SEEK_SET);
-                char send_buf[BUF_SIZE];
-                while (fgets(send_buf, BUF_SIZE, tmp_file) != NULL) {
-                    send(new_fd, send_buf, strlen(send_buf), 0);
-                }
-            }
+        // Allocate a new thread node and insert it into our SLIST.
+        thread_node_t *node = malloc(sizeof(thread_node_t));
+        if (!node) {
+            syslog(LOG_ERR, "Failed to allocate memory for thread node");
+            close(new_fd);
+            continue;
         }
-
-        // Log connection closure
-        syslog(LOG_INFO, "Closed connection from %s:%d", client_ip, client_port);
-
-        // Close the connection
-        close(new_fd);
+        node->client_fd = new_fd;
+        SLIST_INSERT_HEAD(&thread_list_head, node, entries);
+        pthread_create(&node->thread, NULL, connection_handler, node);
```

Comment:
Missing cleanup/free of this memory and the threads created for the list.  These are shown in the valgrind output.